base58/src/error.rs

```diff
@@ -49,12 +85,34 @@ impl std::error::Error for Error {
         use Error::*;
 
         match self {
-            BadByte(_)
-            | BadChecksum(_, _)
+            Decode(ref e) => Some(e),
+            BadChecksum(_, _)
             | InvalidLength(_)
             | InvalidExtendedKeyVersion(_)
             | InvalidAddressVersion(_)
             | TooShort(_) => None,
         }
     }
 }
+
+impl From<DecodeError> for Error {
+    #[inline]
+    fn from(e: DecodeError) -> Self { Self::Decode(e) }
+}
+
+/// Found a invalid byte while decoding base58 string.
+#[derive(Debug, Clone, PartialEq, Eq)]
+#[non_exhaustive]
+pub struct InvalidCharError {
+    /// The invalid byte.
+    pub invalid: u8,
```

Comment:
Accessor?

Reply:
You mean instead of `pub`? I thought we came to consensus that pub field was better than accessor?

Reply:
> You mean instead of `pub`? I thought we came to consensus that pub field was better than accessor?

I recall the opposite conclusion.

Reply:
Agh! damn. And I claim to my children that my brain does actually work just only when I'm programming ...

I've amended my mental model and added "add accessors and make fields private" to my list of error handling chores.

Reply:
To clarify, for struct error types (unless we are 100% sure nothing will change):

- Private fields
- Accessor functions for all relevant fields
- Optional private constructors
- `non_exhaustive` if struct is empty.
- Still do be decided on: Do not re-export from `mod.rs` if hider error

Reply:
We probably want to break this out into a separate discussion thread. This roughly matches my intuition. But I think there are actually two kinds of errors:

* "Leaf errors" which should be structures with private fields and accessors for data that describe the error information (these don't necessarily need to correspond 1-1 to the private fields)
* "Composite errors" which are `#[non_exhaustive]` enums that collect other error types

Reply:
I remember there were a few specific instances where I couldn't imagine changing the layout so I recommended just `pub` fields. If we don't want to deal with exceptions like that I'm fine with doing accessors everywhere.

@apoelstra I have some changes in mind that would kinda mix it up but I got stuck and didn't have brain power and to resolve it. Basically I want to have `ParseError { input: InputString, kind: ParseErrorKind }` in `amount`. To enable nice reporting of spans and multiple errors (later).

Reply:
Ah, yeah, "ErrorKind + context info" is another kind of error that we would probably have.

>I remember there were a few specific instances where I couldn't imagine changing the layout so I recommended just pub fields. If we don't want to deal with exceptions like that I'm fine with doing accessors everywhere.

My vote is to just always use accessors. They are a tighter API anyway because they ensure the fields are read-only and that the error can't be constructed directly.